plaso/parsers/powershell_transcript.py

```diff
@@ -0,0 +1,214 @@
+      # since the system language changes the description of the header fields
+      # we have to go for the line numbers and regex match the timestamp
+      if timestamp_pattern.search(transcript[2]):
+        timestamp_string = timestamp_pattern.search(transcript[2])[1]
+        full_headers = True
+      elif timestamp_pattern.search(transcript[1]):
+        timestamp_string = timestamp_pattern.search(transcript[1])[1]
+        full_headers = False
+      else:
+        parser_mediator.ProduceExtractionWarning('could not find timestamp in '
+          'transcript {0!s} - skipping malformed transcript'.format(transcript))
+        continue
+      # Timestamp format is YYYYMMDDHHmmss
+      time_elements_tuple = int(timestamp_string[:4]),\
+                            int(timestamp_string[4:6]),\
+                            int(timestamp_string[6:8]),\
+                            int(timestamp_string[8:10]),\
+                            int(timestamp_string[10:12]),\
+                            int(timestamp_string[12:])
+      try:
+        start_time = dfdatetime_time_elements.TimeElements(
+          time_elements_tuple=time_elements_tuple)
+      except ValueError:
+        parser_mediator.ProduceExtractionWarning('timestamp \"{0!s}\" seems '
+          'invalid - skipping malformed transcript'.format(time_elements_tuple))
+        continue
+      start_time.is_local_time = True
+      event = time_events.DateTimeValuesEvent(start_time,
+        definitions.TIME_DESCRIPTION_START)
+      # we don't use line breaks here since they aren't displayed nicely in TS
+      if full_headers:
+        command = '; '.join(transcript[18:len(transcript)])
+      else:
+        command = '; '.join(transcript[3:len(transcript)])
+      event_data.command = command
+      # finally create event if values have been set
+      if (event is not None and event_data.command is not None and
+            event_data.command != ''):
+        parser_mediator.ProduceEventWithEventData(event, event_data)
+      else:
+        parser_mediator.ProduceExtractionWarning('skipping transcript {0!s} - '
+          'since relevant event values could not be set'.format(transcript))
+
+
+  def ParseFileObject(self, parser_mediator, file_object):
+    """Parses an PowerShell transcript file-like object.
+
+    Args:
+      parser_mediator (ParserMediator): mediates interactions between parsers
+          and other components, such as storage and dfvfs.
+      file_object (dfvfs.FileIO): file-like object.
+
+    Raises:
+      WrongParser: when the file cannot be parsed.
+    """
+
+    data = file_object.read(self._HEADER_READ_SIZE)
+    if not data == b'\xef\xbb\xbf**********************':
+      raise errors.WrongParser('Not a PowerShell transcript file')
+
+    # The current offset of the file-like object needs to point at
+    # the start of the file to parse the data correctly.
+    file_object.seek(0, os.SEEK_SET)
+    file_bytes = file_object.read()
```

Comment:
@FabFaeb Could this be a problem if the file is quite large in size? Would it be better to read the file in chunks or using a memory-mapped object?

Reply:
Personally, I have never seen a transcript file that was larger than a few MBs - but I can see your point.

I have thought about the reading in chunks, but I found that the unpredictable log format is again a problem. There are no save "boundaries" on which I could separate the chunks. So, very likely, part 1 of a transcript would be read in chunk 1 and part 2 in chunk 2. This would make the parsing of the file into separate "transcripts" (which are then used to generate the events) a whole lot more complicated.

The file_object that is handed to this method is a `dfvfs.FileIO` type. I have not been able to find an example how I could use this as mmapped object. Do you have any reference/example how I could integrate this here?

The simplest (although not the most beautiful) fix I could think of would be to check the size of the file_object and just stop parsing if it is larger than XXX MB - just to have a fail-safe in place.

What would be your favorite approach here?

Reply:
Unfortunately, I think the ideal fix here is more structural - inheriting from PyParsing multi- or single- line parser, and using the existing machinery we have for handling text files in general.

@FabFaeb Could you take a look at that and see what you think? You could take a look at the SCCM parser for some inspiration

Reply:
I had a few minutes, so I started poking at what the pyparsing grammar might look like - https://gist.github.com/Onager/b0162b2c1753e88ffa91923c8a254315 might also provide some inspiration